### PR TITLE
[IOTDB-2685] Create and start config node

### PR DESCRIPTION
**Pull request by CRZbulabula**
Changes:

- ConfigNode start scripts in windows and linux
- ConfigNode configuration checks when start
- Basic JMX settings in ConfigNode

### CODE REVIEW

File: confignode/pom.xml [hunk 1]
```diff
@@ -46,17 +46,24 @@
...
         </dependency>
         <dependency>
             <groupId>org.jetbrains</groupId>
             <artifactId>annotations</artifactId>
             <version>23.0.0</version>
             <scope>provided</scope>
         </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
```
Comment (wangchao316) [inline confignode/pom.xml:61]: <version>${junit.version}</version>
            <scope>test</scope>

File: confignode/src/main/java/org/apache/iotdb/confignode/conf/ConfigNodeConfCheck.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+  private static final ConfigNodeConf conf = ConfigNodeDescriptor.getInstance().getConf();
+
+  private Properties specialProperties;
+
+  public void checkConfig() throws RepeatConfigurationException, IOException {
+
+    String propsDir = ConfigNodeDescriptor.getInstance().getPropsDir();
+    if (propsDir == null) {
+      // Skip configuration check when developer mode or test mode
+      return;
```
Comment (wangchao316) [inline confignode/src/main/java/org/apache/iotdb/confignode/conf/ConfigNodeConfCheck.java:50]: Read the configuration and check whether the configuration can be shared with the standalone version.

File: confignode/src/test/java/org/apache/iotdb/confignode/conf/ConfigNodeDescriptorTest.java [hunk 3]
```diff
@@ -1,76 +0,0 @@
...
-  private final String confPath = System.getProperty(ConfigNodeConf.CONF_NAME, null);
-
-  @Before
-  public void init() {
-    org.apache.catalina.webresources.TomcatURLStreamHandlerFactory.getInstance();
-  }
-
-  @After
-  public void clear() {
-    if (confPath != null) {
```
Comment (wangchao316) [inline confignode/src/test/java/org/apache/iotdb/confignode/conf/ConfigNodeDescriptorTest.java:39]: why does delete test?

File: confignode/src/main/java/org/apache/iotdb/confignode/conf/ConfigNodeConstant.java [hunk 4]
```diff
@@ -20,11 +20,31 @@
...
+
+  public static final String ENV_FILE_NAME = "confignode-env";
+  public static final String CONF_NAME = "iotdb-confignode.properties";
+  public static final String SPECIAL_CONF_NAME = "iotdb-confignode-special.properties";
+
+  public static final String CONFIGNODE_PACKAGE = "org.apache.iotdb.confignode.service";
+  public static final String JMX_TYPE = "type";
+  public static final String CONFIGNODE_JMX_PORT = "confignode.jmx.port";
+
+  public static final String DATA_DIR = "data";
```
Comment (wangchao316) [inline confignode/src/main/java/org/apache/iotdb/confignode/conf/ConfigNodeConstant.java:42]: Add the config_data_dir parameter to the configuration item to support mounting of multiple disks.
 see :"DirectoryManager" in server

File: confignode/src/main/java/org/apache/iotdb/confignode/conf/ConfigNodeConfCheck.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        return;
+      } else {
+        LOGGER.error(
+            "Can't create special configuration file {} for ConfigNode. IoTDB-ConfigNode is shutdown.",
+            specialPropertiesFile.getAbsolutePath());
+        System.exit(-1);
+      }
+    }
+
+    FileInputStream inputStream = new FileInputStream(specialPropertiesFile);
```
Comment (wangchao316) [inline confignode/src/main/java/org/apache/iotdb/confignode/conf/ConfigNodeConfCheck.java:71]: closed after stream use, you could use "try() {}"

File: confignode/src/main/java/org/apache/iotdb/confignode/service/register/ServiceType.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.confignode.service.register;
+
+public enum ServiceType {
+  JMX_SERVICE("JMX ServerService", "JMX ServerService"),
+  CLUSTER_RPC_SERVICE("Cluster RPC Service", "ClusterRPCService");
```
Comment (qiaojialin) [inline confignode/src/main/java/org/apache/iotdb/confignode/service/register/ServiceType.java:23]:
```suggestion
  CLUSTER_RPC_SERVER("Cluster RPC Server", "ClusterRPCServer");
```

File: confignode/src/main/java/org/apache/iotdb/confignode/service/startup/StartupChecks.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+          throw new StartupException(
+              String.format(
+                  "Requires JDK version >= %d, current version is %d",
+                  ConfigNodeConstant.MIN_SUPPORTED_JDK_VERSION, version));
+        } else {
+          LOGGER.info("JDK version is {}.", version);
+        }
+      };
+  private final List<StartupCheck> preChecks = new ArrayList<>();
+  private final List<StartupCheck> defaultTests = new ArrayList<>();
```
Comment (qiaojialin) [inline confignode/src/main/java/org/apache/iotdb/confignode/service/startup/StartupChecks.java:60]:
```suggestion
  private final List<StartupCheck> defaultChecks = new ArrayList<>();
```

File: iotdb-commons/pom.xml [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    specific language governing permissions and limitations
+    under the License.
+
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.apache.iotdb</groupId>
+        <artifactId>iotdb-parent</artifactId>
+        <version>0.13.0-SNAPSHOT</version>
```
Comment (qiaojialin) [inline iotdb-commons/pom.xml:27]:
```suggestion
        <version>0.14.0-SNAPSHOT</version>
```
or use parent version

File: iotdb-commons/pom.xml [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+    <artifactId>iotdb-commons</artifactId>
+    <name>IoTDB Cluster common tools</name>
```
Comment (qiaojialin) [inline iotdb-commons/pom.xml:31]:
```suggestion
    <name>IoTDB commons</name>
```

File: confignode/src/test/java/org/apache/iotdb/confignode/conf/ConfigNodeDescriptorTest.java (hunk 3 above)
Comment (CRZbulabula) [reply confignode/src/test/java/org/apache/iotdb/confignode/conf/ConfigNodeDescriptorTest.java:39]: I found this test is useless...

File: iotdb-commons/pom.xml (hunk 8 above)
Comment (CRZbulabula) [reply iotdb-commons/pom.xml:27]: But parent is 0.13.0 now...

File: iotdb-commons/pom.xml (hunk 9 above)
Comment (CRZbulabula) [reply iotdb-commons/pom.xml:31]: But parent is 0.13.0 now...

File: iotdb-commons/pom.xml (hunk 8 above)
Comment (CRZbulabula) [reply iotdb-commons/pom.xml:27]: But parent is 0.13.0 now...

**Review by qiaojialin — APPROVED** [review]

File: confignode/src/main/java/org/apache/iotdb/confignode/exception/startup/StartupException.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.confignode.exception.startup;
+
+import org.apache.iotdb.confignode.exception.ConfigNodeException;
+
+/** Throws when there exists errors when startup checks */
+public class StartupException extends ConfigNodeException {
```
Comment (OneSizeFitsQuorum) [inline confignode/src/main/java/org/apache/iotdb/confignode/exception/startup/StartupException.java:24]: We may be able to reuse these exceptions with server module, but I'm ok with not dealing with them in this PR

File: confignode/src/main/java/org/apache/iotdb/confignode/service/register/JMXService.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+import javax.management.InstanceNotFoundException;
+import javax.management.MBeanRegistrationException;
+import javax.management.MBeanServer;
+import javax.management.MalformedObjectNameException;
+import javax.management.NotCompliantMBeanException;
+import javax.management.ObjectName;
+
+import java.lang.management.ManagementFactory;
+
+public class JMXService implements IService {
```
Comment (OneSizeFitsQuorum) [inline confignode/src/main/java/org/apache/iotdb/confignode/service/register/JMXService.java:36]: same as IService

File: confignode/src/main/java/org/apache/iotdb/confignode/conf/ConfigNodeConfCheck.java [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.Objects;
+import java.util.Properties;
+
+/**
+ * ConfigNodeConfCheck checks parameters in iotdb-confignode.properties when started, and the
+ * consistency of some specific parameters on restart
+ */
+public class ConfigNodeConfCheck {
```
Comment (OneSizeFitsQuorum) [inline confignode/src/main/java/org/apache/iotdb/confignode/conf/ConfigNodeConfCheck.java:37]: Add TODO for reuse

File: confignode/src/main/java/org/apache/iotdb/confignode/service/register/IService.java [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.confignode.service.register;
+
+import org.apache.iotdb.confignode.exception.startup.StartupException;
+
+/** Any services that run in ConfigNode should implement IService */
+public interface IService {
```
Comment (OneSizeFitsQuorum) [inline confignode/src/main/java/org/apache/iotdb/confignode/service/register/IService.java:24]: Add TODO for reuse

**Review by wangchao316 — APPROVED** [review]

File: confignode/src/main/java/org/apache/iotdb/confignode/service/register/RegisterManager.java [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/** Manager any services that run in ConfigNode */
+public class RegisterManager {
```
Comment (OneSizeFitsQuorum) [inline confignode/src/main/java/org/apache/iotdb/confignode/service/register/RegisterManager.java:31]: Add TODO for reuse

File: confignode/src/main/java/org/apache/iotdb/confignode/utils/ConfigNodeEnvironmentUtils.java [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+
+import static org.junit.Assert.fail;
+
+/** Test environment for ConfigNode UT and IT */
+public class ConfigNodeEnvironmentUtils {
```
Comment (OneSizeFitsQuorum) [inline confignode/src/main/java/org/apache/iotdb/confignode/utils/ConfigNodeEnvironmentUtils.java:35]: Why not put it under the test package?

**Review by OneSizeFitsQuorum — COMMENTED** [review]
PTAL

**Review by OneSizeFitsQuorum — COMMENTED** [review]
PTAL

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Pull request merged**